### PR TITLE
feat: service staking dispenser layout

### PR DESCRIPTION
**Pull request by kupermind**
- Service staking dispenser layout.

### CODE REVIEW

File: contracts/Dispenser.sol [hunk 1]
```diff
@@ -1,10 +1,90 @@
...
+        uint32 numNewOwners;
+        // Epoch end timestamp
+        // 2^32 - 1 gives 136+ years counted in seconds starting from the year 1970, which is safe until the year of 2106
+        uint32 endTime;
+        // Parameters for rewards and top-ups (in percentage)
+        // Each of these numbers cannot be practically bigger than 100 as they sum up to 100%
+        // treasuryFraction + rewardComponentFraction + rewardAgentFraction = 100%
+        // Treasury fraction
+        uint8 rewardTreasuryFraction;
+        // maxBondFraction + topUpComponentFraction + topUpAgentFraction <= 100%
```
Comment (mariapiamo) [inline contracts/Dispenser.sol:40]: I know that we imported this as is now on tokenomics, but maybe we should update the comment saying that maxBondFraction + topUpComponentFraction + topUpAgentFraction <= 100%-serviceStakingFraction

File: contracts/Dispenser.sol [hunk 2]
```diff
@@ -1,10 +1,90 @@
...
+        uint256 transferAmount) external;
+}
+
+interface IServiceStaking {
+    function deposit(uint256 stakingAmount) external;
+}
+
+/// @dev L2 chain Id is not supported.
+/// @param chainId L2 chain Id.
+error L2ChainIdNotSupported(uint256 chainId);
```
Comment (mariapiamo) [inline contracts/Dispenser.sol:81]: L2ChainIdNotSupported is a bit confusing, we can say L2ChainIdOutOfRange

File: contracts/Dispenser.sol [hunk 3]
```diff
@@ -31,6 +140,8 @@ contract Dispenser is IErrorsTokenomics {
     {
         owner = msg.sender;
         _locked = 1;
+        // TODO Define final behavior before deployment
+        paused = Pause.DevIncentivesPaused;
```
Comment (mariapiamo) [inline contracts/Dispenser.sol:144]: why not having a setter also for VoteWeighting?

File: contracts/Dispenser.sol [hunk 4]
```diff
@@ -79,6 +190,14 @@ contract Dispenser is IErrorsTokenomics {
...
     }
 
+    function setRemainingServiceStakingAmount(uint256 epochNumber, uint256 amount) external {
+        // Check for the tokenomics
+        if (msg.sender != tokenomics) {
+            revert ManagerOnly(msg.sender, tokenomics);
+        }
+        mapEpochRemainingServiceStakingAmounts[epochNumber] = amount;
+    }
+
```
Comment (mariapiamo) [inline contracts/Dispenser.sol:200]: Add a notice to claimOwnerIncentives() function saying that it is pausable independently of Treasury pausabilty

File: contracts/Dispenser.sol [hunk 5]
```diff
@@ -113,4 +237,373 @@ contract Dispenser is IErrorsTokenomics {
...
+        uint256 chainId,
+        address stakingTarget,
+        uint256 stakingAmount,
+        bytes memory bridgePayload,
+        uint256 transferAmount
+    ) internal {
+        if (chainId == 1) {
+            // TODO Inject factory verification here
+            // TODO Check for the numEpochs(Tokenomics) * rewardsPerSecond * numServices * epochLength(Tokenomics)
+            // Get hash of target.code, check if the hash is present in the registered factory
```
Comment (mariapiamo) [inline contracts/Dispenser.sol:251]:
```suggestion

```

Comment (mariapiamo) [reply contracts/Dispenser.sol:251]: This is not necessary since the verifier logic in the factory

File: contracts/Dispenser.sol [hunk 6]
```diff
@@ -113,4 +237,373 @@ contract Dispenser is IErrorsTokenomics {
...
+            // Approve the OLAS amount for the staking target
+            IToken(olas).approve(stakingTarget, stakingAmount);
+            IServiceStaking(stakingTarget).deposit(stakingAmount);
+            // bridgePayload is ignored
+        } else {
+            address depositProcessor = mapChainIdTargetProcessors[chainId];
```
Comment (mariapiamo) [inline contracts/Dispenser.sol:257]: we need to check that depositProcessor was defined, so we need to revert if ```depositProcessor=address(0)```

File: contracts/Dispenser.sol [hunk 7]
```diff
@@ -113,4 +237,373 @@ contract Dispenser is IErrorsTokenomics {
...
+            // TODO: mint directly or mint to dispenser and approve one by one?
```
Comment (mariapiamo) [inline contracts/Dispenser.sol:258]: we believe we need to mint to the dispenser because is simple

File: contracts/Dispenser.sol [hunk 8]
```diff
@@ -113,4 +237,373 @@ contract Dispenser is IErrorsTokenomics {
...
+            // Approve the OLAS amount for the staking target
```
Comment (mariapiamo) [inline contracts/Dispenser.sol:259]:
```suggestion

```

Comment (mariapiamo) [reply contracts/Dispenser.sol:259]: just a simple transfer, this comment is not needed

File: contracts/Dispenser.sol [hunk 9]
```diff
@@ -113,4 +237,373 @@ contract Dispenser is IErrorsTokenomics {
...
+            // TODO: optimize not to read several times in a row same epoch info
+            // Get service staking info
+            ITokenomicsInfo.ServiceStakingPoint memory serviceStakingPoint =
+                ITokenomicsInfo(tokenomics).mapEpochServiceStakingPoints(j);
+
+            ITokenomicsInfo.EpochPoint memory ep = ITokenomicsInfo(tokenomics).mapEpochTokenomics(j);
+            uint256 endTime = ep.endTime;
+
+            // Get the staking weight for each epoch and the total weight
+            // TODO math from where we need to get the weight - endTime or endTime + WEEK, seems like the latter because values are written for the nextTime
```
Comment (mariapiamo) [inline contracts/Dispenser.sol:344]: let's sync about this, we are not convinced that we need endTime +week

File: contracts/Dispenser.sol [hunk 10]
```diff
@@ -113,4 +237,373 @@ contract Dispenser is IErrorsTokenomics {
...
+                totalAmountReturn += returnAmount;
+            } else {
+                // Otherwise, allocate staking amount to corresponding contracts
+                stakingAmount = (availableStakingAmount * stakingWeight) / 1e18;
+                if (stakingAmount > serviceStakingPoint.maxServiceStakingAmount) {
+                    // Adjust the return amount
+                    returnAmount = stakingAmount - serviceStakingPoint.maxServiceStakingAmount;
+                    totalAmountReturn += returnAmount;
+                    // Adjust the staking amount
+                    stakingAmount = serviceStakingPoint.maxServiceStakingAmount;
```
Comment (mariapiamo) [inline contracts/Dispenser.sol:369]: add an event saying that the target received a very low weight, so it does not qualify for rewards

File: contracts/Dispenser.sol [hunk 11]
```diff
@@ -113,4 +237,373 @@ contract Dispenser is IErrorsTokenomics {
...
+                }
+                totalStakingAmount += stakingAmount;
+            }
+
+            // Adjust remaining service staking amounts
+            uint256 remainingServiceStakingAmount = mapEpochRemainingServiceStakingAmounts[j];
+            // Combine staking and return amounts as they both need to be subtracted from the remaining amount
+            stakingAmount += returnAmount;
```
Comment (mariapiamo) [inline contracts/Dispenser.sol:377]: add an event saying that the target contract can receive at most maxServiceStakingAmount OLAS per epoch

File: contracts/Dispenser.sol [hunk 12]
```diff
@@ -113,4 +237,373 @@ contract Dispenser is IErrorsTokenomics {
...
+
+        // Dispense to a service staking target
+        _distribute(chainId, target, stakingAmount, bridgePayload, transferAmount);
+
+        // TODO: Tokenomics - return totalReturnAmount into EffectiveSatking (or another additional variable tracking returns to redistribute further)
+        if (returnAmount > 0) {
+            ITokenomicsInfo(tokenomics).refundFromServiceStaking(returnAmount);
+        }
+
+        emit ServiceStakingIncentivesClaimed(msg.sender, stakingAmount, returnAmount);
```
Comment (mariapiamo) [inline contracts/Dispenser.sol:436]: can we also emit the `target` and `chainId`?

File: contracts/Dispenser.sol [hunk 13]
```diff
@@ -113,4 +237,373 @@ contract Dispenser is IErrorsTokenomics {
...
+        if (totalReturnAmount > 0) {
+            ITokenomicsInfo(tokenomics).refundFromServiceStaking(totalReturnAmount);
+        }
+
+        emit ServiceStakingIncentivesClaimed(msg.sender, totalStakingAmount, totalReturnAmount);
+
+        _locked = 1;
+    }
+
+    /// @dev Sets deposit processor contracts addresses and L2 chain Ids.
```
Comment (mariapiamo) [inline contracts/Dispenser.sol:542]:
```suggestion
    /// @dev Sets deposit processor contracts addresses and L2 chain Ids.
    /// @notice If there is a need to stop processing a specific L2 chain Id, set depositProcessors=0
    /// @notice If there is a need to update processing of specific L2 chain Id, set the new depositProcessors address.
```

File: contracts/staking/ArbitrumDepositProcessorL1.sol [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+        // Send a message to the staking dispenser contract on L2 to reflect the transferred OLAS amount
+        //cost = maxSubmissionCost + maxGas * gasPriceBid;
+        uint256 sequence = IBridge(l1MessageRelayer).createRetryableTicket{value: cost}(l2TargetDispenser, 0,
+            maxSubmissionCost, refundAccount, refundAccount, maxGas, gasPriceBid, data);
+
+        emit MessageSent(sequence, targets, stakingAmounts, transferAmount);
+    }
+
+    /// @dev Process message received from L2.
+    /// @param data Bytes message data sent from L2.
```
Comment (mariapiamo) [inline contracts/staking/ArbitrumDepositProcessorL1.sol:184]: re-check that this is automatically called after challenge period

File: contracts/staking/DefaultDepositProcessorL1.sol [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+    /// @param _l1MessageRelayer L1 message relayer bridging contract address.
+    /// @param _l2TargetChainId L2 target chain Id.
+    constructor(
+        address _olas,
+        address _l1Dispenser,
+        address _l1TokenRelayer,
+        address _l1MessageRelayer,
+        uint256 _l2TargetChainId
+    ) {
+        if (_l1Dispenser == address(0) || _l1TokenRelayer == address(0) || _l1MessageRelayer == address(0)) {
```
Comment (mariapiamo) [inline contracts/staking/DefaultDepositProcessorL1.sol:48]: revert if ```_olas == address(0)```

File: contracts/staking/DefaultTargetDispenserL2.sol [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+        // Check for the contract ownership
+        if (msg.sender != owner) {
+            revert OwnerOnly(msg.sender, owner);
+        }
+
+        paused = 1;
+        emit Unpaused();
+    }
+
+    receive() external payable {}
```
Comment (mariapiamo) [inline contracts/staking/DefaultTargetDispenserL2.sol:268]: Is this method needed for gas fee?

File: contracts/staking/DefaultTargetDispenserL2.sol [hunk 17]
```diff
@@ -0,0 +1,269 @@
+// SPDX-License-Identifier: MIT
```
Comment (mariapiamo) [inline contracts/staking/DefaultTargetDispenserL2.sol:1]: natspec is incomplete here

File: contracts/staking/DefaultTargetDispenserL2.sol [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+    event ServiceStakingParametersUpdated(uint256 rewardsPerSecondLimit);
+    event MessageSent(uint256 indexed sequence, address indexed messageSender, address indexed l1Processor, uint256 amount);
+    event MessageReceived(address indexed messageSender, uint256 chainId, bytes data);
+    event WithheldAmountSynced(address indexed sender, uint256 amount);
+    event Paused();
+    event Unpaused();
+
+    // Gas limit for sending a message to L1
+    uint256 public constant GAS_LIMIT = 100_000;
+    // OLAS address
```
Comment (mariapiamo) [inline contracts/staking/DefaultTargetDispenserL2.sol:35]:
```suggestion
    // L2 OLAS address
```

File: contracts/staking/DefaultTargetDispenserL2.sol [hunk 19]
```diff
@@ -0,0 +1,269 @@
...
+    address public immutable olas;
+    // Proxy factory address
+    address public immutable proxyFactory;
```
Comment (mariapiamo) [inline contracts/staking/DefaultTargetDispenserL2.sol:38]: why not having poxyFactory mutable with a setter?

File: contracts/staking/DefaultTargetDispenserL2.sol [hunk 20]
```diff
@@ -0,0 +1,269 @@
...
+    // Source processor address on L1 that is authorized to propagate the transaction execution across the bridge
+    address public immutable l1SourceProcessor;
+    // Source processor chain Id
+    uint256 public immutable l1SourceChainId;
+    // Amount of OLAS withheld due to service staking target invalidity
+    uint256 public withheldAmount;
+    // Nonce for each staking batch
+    uint256 public stakingBatchNonce;
+    // rewardsPerSecondLimit
+    uint256 public rewardsPerSecondLimit;
```
Comment (mariapiamo) [inline contracts/staking/DefaultTargetDispenserL2.sol:52]:
```suggestion

```

File: contracts/staking/DefaultTargetDispenserL2.sol [hunk 21]
```diff
@@ -0,0 +1,269 @@
...
+    address public immutable l2MessageRelayer;
+    // Source processor address on L1 that is authorized to propagate the transaction execution across the bridge
+    address public immutable l1SourceProcessor;
+    // Source processor chain Id
+    uint256 public immutable l1SourceChainId;
+    // Amount of OLAS withheld due to service staking target invalidity
+    uint256 public withheldAmount;
+    // Nonce for each staking batch
+    uint256 public stakingBatchNonce;
+    // rewardsPerSecondLimit
```
Comment (mariapiamo) [inline contracts/staking/DefaultTargetDispenserL2.sol:51]:
```suggestion

```

Comment (mariapiamo) [reply contracts/staking/DefaultTargetDispenserL2.sol:51]: rewardsPerSecondLimit was removed after the introduction of the verifier logic in the factory

File: contracts/staking/DefaultTargetDispenserL2.sol [hunk 22]
```diff
@@ -0,0 +1,269 @@
...
+
+    // Process the data
+    function _processData(bytes memory data) internal {
+        (address[] memory targets, uint256[] memory amounts) = abi.decode(data, (address[], uint256[]));
+
+        uint256 batchNonce = stakingBatchNonce;
+        for (uint256 i = 0; i < targets.length; ++i) {
+            address target = targets[i];
+            uint256 amount = amounts[i];
+            // TODO Shall we account for paused here and just queue, if paused?
```
Comment (mariapiamo) [inline contracts/staking/DefaultTargetDispenserL2.sol:95]: discuss this later, if paused send all the amoun in withheld?

File: contracts/staking/DefaultTargetDispenserL2.sol [hunk 23]
```diff
@@ -0,0 +1,269 @@
...
+                if (success) {
+                    // Approve and transfer OLAS to the service staking target
+                    IToken(olas).approve(target, amount);
+                    IServiceStaking(target).deposit(amount);
+                    emit ServiceStakingTargetDeposited(target, amount);
+
+                } else {
+                    // Withhold OLAS for further usage
+                    withheldAmount += amount;
+                    emit ServiceStakingAmountWithheld(target, amount);
```
Comment (mariapiamo) [inline contracts/staking/DefaultTargetDispenserL2.sol:117]: can we also emit withheldAmount?

File: contracts/Tokenomics.sol [hunk 24]
```diff
@@ -228,6 +256,13 @@ contract Tokenomics is TokenomicsConstants, IErrorsTokenomics {
     // Mapping of component / agent Id => incentive balances
     mapping(uint256 => mapping(uint256 => IncentiveBalances)) public mapUnitIncentives;
 
+    // TODO: maxServiceStaking for each epoch?
+    // effectiveServiceStaking
+    // This number cannot be practically bigger than the inflation remainder of OLAS
+    uint96 public effectiveServiceStaking;
```
Comment (mariapiamo) [inline contracts/Tokenomics.sol:262]: we just use ```serviceStakingAmount```, so this can be removed

**Review by mariapiamo — APPROVED** [review]

**Review by 77ph — APPROVED** [review]

**Review by mariapiamo — APPROVED** [review]

**Pull request merged**